### ISSUE TITLE
Sub socket silently ignores bulk subscriptions

### ISSUE THREAD

**Issue by derSeddy** [zeromq/libzmq#4064]
# Issue description

When subscribing to channels in bulk the subscription is not successful and messages are not received. I'm unsure if the subscription message is dropped by the sub socket or ignored by the pub socket. 

This bug might be related or even the cause of https://github.com/zeromq/libzmq/issues/4063.

# Environment

* libzmq version (commit hash if unreleased): 
Tested with `4.3.1` and current master `97b5f8560d9ce561949537b6c4c98e8eb2138148`
* OS: 
Debian 4.19.118-2+deb10u1 (2020-06-07) x86_64 GNU/Linux

# Minimal test code / Steps to reproduce the issue

In a for loop the code tries several numbers of subscriptions, from 100 to 1500 channels, and send a message on each channel which must be received to continue.

Code includes three test cases.
1: Every subscription is done by itself and right before the message is received
2: All subscriptions are done in bulk but to work around the problem a dummy receive is called after every subscription
3: All subscriptions are done in bulk, on my hardware this deadlocks at 1000 subscriptions

```cpp
#include <atomic>
#include <iostream>
#include <thread>
#include <unistd.h>
#include <zmq.hpp>

int main()
{
	const std::string address("tcp://127.0.0.1:58765");
	zmq::context_t context;

	std::atomic<size_t> atCount(0);

	auto channelName = [&]()
	{
		return "chan_" + std::to_string(atCount) + "_";
	};

	std::thread sender([&]()
	{
		zmq::socket_t pub(context, ZMQ_PUB);
		pub.bind(address);
		while(true)
		{
			const std::string channelSnd = channelName();
			pub.send(channelSnd.data(), channelSnd.size(), ZMQ_SNDMORE);
			pub.send(channelSnd.data(), channelSnd.size());
			usleep(1000);
		}
	});



	auto waitForRecv = [](zmq::socket_t& sub, const std::string& channelSnd)
	{
		while(true)
		{
			zmq::message_t msg;
			if(sub.recv(&msg))
			{
				const std::string channelRcv((const char*)msg.data(), msg.size());
				sub.recv(&msg);
				const std::string data((const char*)msg.data(), msg.size());
				if(channelRcv == channelSnd && channelRcv == data)
					return;
			}
		}
	};

	{
		for(size_t subCount = 100; subCount <= 1500; subCount += 100)
		{
			zmq::socket_t sub(context, ZMQ_SUB);
			sub.connect(address);

			// subscribe and receive single channels
			for(atCount = 0; atCount < subCount; ++atCount)
			{
				const std::string channelSnd = channelName();
				sub.setsockopt(ZMQ_SUBSCRIBE, channelSnd.data(), channelSnd.size());
				waitForRecv(sub, channelSnd);
			}

			std::cout << "Single success for " << subCount << std::endl;
		}
	}

	{
		for(size_t subCount = 100; subCount <= 1500; subCount += 100)
		{
			zmq::socket_t sub(context, ZMQ_SUB);
			sub.connect(address);

			// bulk subscribe
			for(atCount = 0; atCount < subCount; ++atCount)
			{
				const std::string channel = channelName();
				sub.setsockopt(ZMQ_SUBSCRIBE, channel.data(), channel.size());
				zmq::message_t msg;
				sub.recv(&msg, ZMQ_DONTWAIT);
			}

			// receive single channels
			for(atCount = 0; atCount < subCount; ++atCount)
				waitForRecv(sub, channelName());

			std::cout << "Bulk workaround success for " << subCount << std::endl;
		}
	}

	{
		for(size_t subCount = 100; subCount <= 1500; subCount += 100)
		{
			zmq::socket_t sub(context, ZMQ_SUB);
			sub.connect(address);

			// bulk subscribe
			for(atCount = 0; atCount < subCount; ++atCount)
			{
				const std::string channel = channelName();
				sub.setsockopt(ZMQ_SUBSCRIBE, channel.data(), channel.size());
				zmq::message_t msg;
			}

			// receive single channels
			for(atCount = 0; atCount < subCount; ++atCount)
				waitForRecv(sub, channelName());

			std::cout << "Bulk success for " << subCount << std::endl;
		}
	}
}
```

Compiled and run with
`g++ main.cpp -lzmq -std=c++14 -pthread && ./a.out`
or
`clang main.cpp -lzmq -std=c++14 -pthread -lstdc++ && ./a.out`.

# What's the actual result? (include assertion message & call stack if applicable)

Test freezes before reaching the end:
```
Single success for 100
Single success for 200
Single success for 300
Single success for 400
Single success for 500
Single success for 600
Single success for 700
Single success for 800
Single success for 900
Single success for 1000
Single success for 1100
Single success for 1200
Single success for 1300
Single success for 1400
Single success for 1500
Bulk workaround success for 100
Bulk workaround success for 200
Bulk workaround success for 300
Bulk workaround success for 400
Bulk workaround success for 500
Bulk workaround success for 600
Bulk workaround success for 700
Bulk workaround success for 800
Bulk workaround success for 900
Bulk workaround success for 1000
Bulk workaround success for 1100
Bulk workaround success for 1200
Bulk workaround success for 1300
Bulk workaround success for 1400
Bulk workaround success for 1500
Bulk success for 100
Bulk success for 200
Bulk success for 300
Bulk success for 400
Bulk success for 500
Bulk success for 600
Bulk success for 700
Bulk success for 800
Bulk success for 900
Bulk success for 1000
```
I never saw the test succeed and force closed it after a few minutes.

# What's the expected result?

The test should run until `Bulk success for 1500` is printed and exit.

**Comment by derSeddy** [comment]
I am aware that the test passes fine when increasing buffer sizes with `setsockopt(ZMQ_SNDHWM, large_number);` but I think the real problem is that this behaviour is ignored silently and no error is thrown or returned on `setsockopt`. It is ok if some messages are lost due to slow connections etc, but the subscription messages should be guarantied to succeed.

**Comment by bluca** [comment]
Subscription messages are messages, and thus are subject to the high water mark. You need to change it accordingly if your volume is higher than the default.

**Comment by pizzard** [comment]
Assume I adjust the buffer according to my normal operation requirements, the system will work in absence of disconnects.
So when I over a larger timeframe slowly increase my number of subscriptions and send more and more messages but stay well within my buffer limits so everything works fine, and then a reconnection event occurs, zmq will not notify about that (which is fine), but then internally resubscribes everything in bulk neglecting its own buffer sizes or any high water marks, so drop all subscription messages and afterwards drop all messages without noticing me or ever recovering (which is not fine, right?). So my system will go from fully working state to completely broken, 90% message drop forever, without any possibility for me to notice and the only possible fix is to make buffers orders of magnitude bigger than the working system would need in normal operation?

**Comment by bluca** [comment]
> the only possible fix is to make buffers orders of magnitude bigger than the working system would need in normal operation?

Yes, that is how it works - subscription messages are messages, and thus are subject to the high water mark

**Comment by pizzard** [comment]
We have done further work on this issue but it is still unsolved. 
The following factors amplify this problem here quite intensively:
* When a connection is lost and reestablished, initially the available bandwidth is usually very low with WiFi, LTE and other links. It then normalizes quickly to its high normal bandwidth. Buffer have filled up during the downtime.
* In this initial reconnect period, zmq under the hood generates all subscription messages and fires them. 
* As the initial bandwidth is insufficient and the buffer are already full, zmq under the hood immediately drops a large part of the subscription messages. They never even attempt to go somewhere.
* zmq subscription messages have no feedback, response or confirmation that the subscription happened of any kind, so neither can client code ever see that the subscription didn't happen nor will zmq ever retry anything or do other steps to recover. 
* The silence heard comes from missing subscriptions, but is completely indistinguishable to silence from the absence of events. 
* canary channels (which are subscribed just to play ping pong and detect the problem) are unreliable, as sometimes their subscription goes through but not others. Also then zmq fails to achieve to alleviate me from session handling. 

Would it be feasible to add a simple subscription confirmation message so the subscription can be retried if it was requested but never went through? 
Would it be feasible to modify the buffers so that subscription messages have priority and are added to the buffers even when they are full?

Am I the only one using zmq over bandwidth-constrained links? Am I missing an obvious solution here?

**Comment by bluca** [comment]
Use XPUB/XSUB
